tests: Update to Ginkgo 2.0

oshoval opened the pull request:
**What this PR does / why we need it:**
Update Kubevirt to use Ginkgo 2.0 in order to be able to use the new features of it.

* Imports were changed to use Ginkgo 2.0 all over.
* `kubevirt/qe-tools` was bumped to v0.1.8 so it will also have Ginkgo 2.0 interface.
* Since Ginkgo 2.0 changed the reporters, and the new built-in reporter has different
XSD scheme, which isn't compatible with prow spyglass, the old reporter
is still used as a custom reporter.
* Handled deprecation warnings / errors.
* Since mock `ctrl.Finish()` is called automatically, it uncovers some unit tests that
need to be updated.
This PR updates the needed calls so it will pass, but an issue [1] is opened to
double check the tests and see if additional actions should be taken.
The redundant `Finish` calls were removed.
* `device-manager` Bazel file was changed manually, because it isn't
changed by `make deps-update`
* Ginkgo installation was removed from kubevirt builder image, as it doesn't affect the build
`KUBEVIRT_BUILDER_IMAGE` was not updated yet.
* `DeprecatedBeforeAll` should be removed on a follow PR.

[1] https://github.com/kubevirt/kubevirt/issues/7332

**Which issue(s) this PR fixes** (optional, in fixes #<issue number>(, fixes #<issue_number>, ...) format, will close the issue(s) when PR gets merged):
Fixes #

**Special notes for your reviewer:**
Based on https://onsi.github.io/ginkgo/MIGRATING_TO_V2
Some of the changes were inspired by https://github.com/gardener/gardener/pull/5313
Once this PR is merged, we can merge also https://github.com/kubevirt/project-infra/pull/1931

**Release note:**

```release-note
The Ginkgo dependency has been upgraded to v2.1.3 (major version upgrade)
```

oshoval commented:
/uncc @alonSadan @ILpinto @rhrazdil @yuhaohaoyu

qinqon commented on pkg/host-disk/host-disk_test.go:
```diff
@@ -200,7 +200,7 @@ var _ = Describe("HostDisk", func() {
 				})
 
-				It("Should NOT subtract reserve if there is enough space on storage for requested size", func(done Done) {
+				It("Should NOT subtract reserve if there is enough space on storage for requested size", func() {
```
This is not following migration guide https://onsi.github.io/ginkgo/MIGRATING_TO_V2#migration-strategy, are we ok with this ? 

This shoul be migrated to
```golang
It("...", func() {
	done := make(chan interface{})
	go func() {
		// user test code to run asynchronously
		close(done) //signifies the code is done
	}()
	Eventually(done, timeout).Should(BeClosed())
})
```

oshoval replied:
All the tests are passing, they are running sync
I think it is just a recommendation for the equivalent behavior of the deprecated feature,
the feature was removed because it adds confusion (according the guide),

I think that it will be better to have one level of parallel that we already have,
ad hoc in case something is needed we can improve, but not globally just because we want to preserve the async behavior which is now officially deprecated

qinqon commented on hack/functests.sh:
```diff
@@ -59,7 +59,7 @@ function functest() {
     fi
 
-    _out/tests/ginkgo -r -slow-spec-threshold=60s $@ _out/tests/tests.test -- -kubeconfig=${kubeconfig} -container-tag=${docker_tag} -container-tag-alt=${docker_tag_alt} -container-prefix=${functest_docker_prefix} -image-prefix-alt=${image_prefix_alt} -oc-path=${oc} -kubectl-path=${kubectl} -gocli-path=${gocli} -installed-namespace=${namespace} -previous-release-tag=${PREVIOUS_RELEASE_TAG} -previ ...
+    _out/tests/ginkgo -timeout=3h -r -slow-spec-threshold=60s $@ _out/tests/tests.test -- -kubeconfig=${kubeconfig} -container-tag=${docker_tag} -container-tag-alt=${docker_tag_alt} -container-prefix=${functest_docker_prefix} -image-prefix-alt=${image_prefix_alt} -oc-path=${oc} -kubectl-path=${kubectl} -gocli-path=${gocli} -installed-namespace=${namespace} -previous-release-tag=${PREVIOUS_RELEASE ...
```
Why we need this now ? maybe we did something wrong and now is not running in parallel ?

qinqon commented on tests/reporter/kubernetes.go:
```diff
@@ -95,16 +100,16 @@ func (r *KubernetesReporter) BeforeSuiteDidRun(_ *types.SetupSummary) {
 }
 
 func (r *KubernetesReporter) SpecWillRun(_ *types.SpecSummary) {
-	fmt.Fprintf(ginkgo.GinkgoWriter, "On failure, artifacts will be collected in %s/%d_*\n", r.artifactsDir, r.failureCount+1)
```
We have to call this code to the `ReportBeforeEach`

oshoval replied:
will print as part of the function called by
https://github.com/kubevirt/kubevirt/pull/7326/commits/2015e0e7073beb63f9599279b0e8751a3b5e4d13#diff-0a9a24f44dfc4029d6c2465ca6246bf66829e97d236da6ff80172a8a9c73866aR115

so it will print only on error, not on every test
ok ?

qinqon commented on tests/tests_suite_test.go:
```diff
@@ -60,31 +60,31 @@ func TestTests(t *testing.T) {
 
 	outputEnricherReporter := reporter.NewCapturedOutputEnricher(
-		ginkgo_reporters.NewJUnitReporter(junitOutput),
+		v1reporter.NewV1JUnitReporter(junitOutput),
```
Do we need the custom junit reporter anymore ?

oshoval replied:
yes, we are using it and not the built-in reporter, which has different XSD scheme
using the new reporter can be done on a follow PR, but needs lot of adjustments,
for the tests, unit-tests, junit-merger, flake finder, prow spyglass and so on
so for now we are using the good old version as a custom reporter (also polarion reporter that is used uses the same)
this PR is too massive already, so this was the most self contained solution i found (tried both and it didn't converge well)

inspired by
https://github.com/gardener/gardener/pull/5313

qinqon reviewed (changes requested):
Thanks for the massive job, added some requests.

oshoval commented:
first push - rebase
2nd push addressed https://github.com/kubevirt/kubevirt/pull/7326#discussion_r822654993

oshoval commented:
change: Updated BeforeSuite usages, (and few more minor changes),
instead of using "Ordered" which forces running tests in serial and takes more time

oshoval commented:
Fixed a unit test broken by the last push
just https://github.com/kubevirt/kubevirt/pull/7326/commits/efd4b17b08cae5671eec37d4723e2310795ebab7 was changed

oshoval commented:
the flakes of https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/7326/pull-kubevirt-e2e-kind-1.22-sriov/1501902070926544896
appear also on other PRs

qinqon commented:
@oshoval can you rebase it ?

oshoval commented:
Rebased

oshoval commented:
/retest 

not related

qinqon commented:
/lgtm
Great job.

dhiller commented on tests/vmi_configuration_test.go:
```diff
@@ -1923,7 +1922,7 @@ var _ = Describe("[sig-compute]Configurations", func() {
 			Expect(err).ToNot(HaveOccurred())
 
 			Eventually(ThisDV(dataVolume), 240).Should(Or(HaveSucceeded(), BeInPhase(cdiv1.WaitForFirstConsumer)))
-		}, 60)
```
I've seen this in a couple other places now that those default timeout values have been removed. Why do we remove the explicit timeout values? I guess there was a reason to define them as they were?

Also they differ widely, I've seen some as small as `15` while the biggest was `720`? So what is the default timeout value now?

I.e. https://github.com/kubevirt/kubevirt/pull/7326/files#diff-d9916fc2bf48043a4899c599755af277e7b8c341d8a737999f7a6667f61596f1L110

oshoval replied:
The timeout decorator is deprecated and not supported anymore,
as well as the async testing feature (with the done channel)
https://onsi.github.io/ginkgo/MIGRATING_TO_V2#removed-async-testing

some of the timeout values might be a legacy copy paste,
felt it is better to ad hoc update tests in case needed instead of updating all the tests with limitation just because there used to be one, (which didn't cause timeouts before as well),

I don't think that  know if there is a per test default timeout now ?
see please https://github.com/golang/go/issues/48157

couldn't open the link you pasted, what is the file / line please ?

dhiller replied:
I don't see any mention that timeout is deprecated, I only see this note here: https://onsi.github.io/ginkgo/MIGRATING_TO_V2#timeout-behavior

There is no statement I can find that the previous timeout behavior without the `done` channel does not exist any more.

oshoval replied:
Having the timeout gives a runtime error, about unknown decorator

oshoval replied:
Having the timeout gives a runtime error, about unknown decorator (`BeforeEach` has new supported decorators)
```
Unknown Decorator
tests/vmi_configuration_test.go:1914
  [BeforeEach] node was passed an unknown decorator: '60'

  Learn more at: http://onsi.github.io/ginkgo/#node-decorators-overview

```

also see here please
https://onsi.github.io/ginkgo/MIGRATING_TO_V2#removed-async-testing
the timeout won't be enforced

dhiller reviewed (commented):
I had only a partial look, I'll take another look again tomorrow.

oshoval commented on hack/builder/Dockerfile:
```diff
@@ -89,7 +89,7 @@ RUN set -x && \
     go get -v -u github.com/rmohr/mock/mockgen@v0.0.0-20170327080805-5980f5d9a70d && \
     go get -v -u github.com/rmohr/go-swagger-utils/swagger-doc@v0.0.0-20181110145135-878f7deb1d17 && \
-    go get -v -u github.com/onsi/ginkgo/ginkgo@v1.12.0 && \
+    go get -v -u github.com/onsi/ginkgo/v2@v2.1.3 && \
```
it seems unused indeed, i also still use the previous kubevirt builder image
i don't think it will fail with the new image, as the new image would simply use v2 
I can try to remove it, just tried to align it

oshoval commented on pkg/ignition/ignition_test.go:
```diff
@@ -24,36 +24,38 @@ import (
 	// const ignitionLocalDir = "/var/run/libvirt/ignition-dir"
 	var vmi *v1.VirtualMachineInstance
 
-	BeforeSuite(func() {
```
BeforeSuite must be called on top level, see please the error otherwise
```
It looks like you are trying to add a [BeforeSuite] node within a container
node.

BeforeSuite can only be called at the top level.

Learn more at:
http://onsi.github.io/ginkgo/#suite-setup-and-cleanup-beforesuite-and-aftersuite
```

oshoval commented:
Changes: 
Added DeferCleanup instead AfterSuite [link](https://github.com/kubevirt/kubevirt/pull/7326#discussion_r825909552)
Removed ginkgo installation from kubevirt builder [link](https://github.com/kubevirt/kubevirt/pull/7326#discussion_r825912242)

oshoval commented:
/test pull-kubevirt-e2e-k8s-1.22-sig-compute-migrations

not related (failure during cluster creation, due to go lib fetching)

oshoval commented:
Change: inline declaration of variable in the BeforeSuite, as it is not used outside of it

oshoval commented:
/retest 

unrelated, some CI jobs were failed today generally due to some unavailable go packages

oshoval commented:
The failures are unrelated

oshoval commented:
rebased

oshoval commented:
/test pull-kubevirt-goveralls  

not related

dhiller reviewed (approved):
/approve

Thank you, great work!

oshoval commented:
Thanks Daniel :)

/cc @qinqon can you please re-review ?

oshoval commented:
/retest

oshoval commented:
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/7326/pull-kubevirt-e2e-kind-1.23-vgpu/1505908123662028800

saw this failure on other PRs as well (also it is after the tests, when trying to delete the kind node)

/retest

qinqon commented:
/lgtm
